hw/misc/hermes.c

```diff
@@ -399,18 +575,62 @@ static uint64_t hermes_bar0_read(void *opaque, hwaddr addr, unsigned size)
             val &= 0xFFFF;
             break;
         }
+    } else if (addr >= HERMES_BAR0_CMD_REQ &&
+               addr + size <= HERMES_BAR0_CMD_REQ +
+               sizeof(struct hermes_cmd) * UBPF_ENGINES) {
+        ptr = (uint32_t *) &((uint8_t *) &bar0_init.commands)
+              [addr - HERMES_BAR0_CMD_REQ];
+        val = *ptr;
+    } else if (addr >= HERMES_BAR0_CMD_CTRL && addr + size
+               <= HERMES_BAR0_CMD_CTRL +
+               sizeof(struct hermes_cmd_ctrl) * UBPF_ENGINES) {
+        ptr = (uint32_t *) &((uint8_t *) &bar0_init.cmdctrl)
+              [addr - HERMES_BAR0_CMD_CTRL];
+        val = *ptr;
     } else {
         hermes_bar_warn_invalid(0, addr);
     }
 
     return val;
 }
 
-/* BAR 0 is read-only */
+/* BAR 0 is partly read-only */
 static void hermes_bar0_write(void *opaque, hwaddr addr, uint64_t val,
                 unsigned size)
 {
-    hermes_bar_warn_read_only(0, addr);
+    int engine, offset;
+    HermesState *hermes = opaque;
+
+    if (addr >= HERMES_BAR0_CMD_REQ && addr < HERMES_BAR0_CMD_CTRL) {
+        engine = (addr - HERMES_BAR0_CMD_REQ) / sizeof(struct hermes_cmd);
+        offset = (addr - HERMES_BAR0_CMD_REQ) % sizeof(struct hermes_cmd);
+
+        if (engine > UBPF_ENGINES - 1) {
+            hermes_bar_warn_invalid(0, addr);
+            return;
+        }
+
+        memcpy((void *) &bar0_init.commands[engine] + offset, &val, size);
+    } else if (addr >= HERMES_BAR0_CMD_CTRL) {
+        engine = (addr - HERMES_BAR0_CMD_CTRL) / sizeof(struct hermes_cmd_ctrl);
```

Comment:
I just realized we have to check the offset here as well (not just the engine). Offset 0 is EHCMDEXEC and is RW, offset 1 is EHCMDDONE and is RO and offsets 2-7 are invalid.